### ISSUE TITLE
Analyzer suggesting MemberData provider use TheoryData type

### ISSUE THREAD

**Issue by bradwilson** [xunit/xunit#1244]
_From @marcind on September 1, 2016 21:17_

The various generic variants of `TheoryData` allow for compile-time checks of theory data type compatibility, suggest using that over `IEnumerable<object[]>`

_Copied from original issue: xunit/xunit.analyzers#8_

**Comment by jamesqo** [comment]
@marcind I do not think this is a good idea. There are lots of codebases out there that use `IEnumerable<object[]>` instead of `TheoryData`, for example [CoreFX](https://github.com/dotnet/corefx).

It might be more sensible to enforce consistency of return type within a certain class-- a test class needs to use either all `IEnumerable<object[]>` or all `TheoryData` for its `MemberData` return types.

**Comment by marcind** [comment]
It is a good idea, because it allows the compiler to enforce that items in the theory data source are correctly type. Combine that with the first part of #1391, and you should be guaranteed that the data matches. Replicating the behavior of the compiler (i.e. the second part of #1391) is a lot harder and possibly not the greatest idea.

**Comment by jamesqo** [comment]
@marcind I see your point. However, I'm guessing there are a lot of projects out there that would be unwilling to make the switch, as a return type of plain `IEnumerable<object[]>` has always been allowed. As long as they can disable the rule, I guess.

**Comment by JamesTerwilliger** [comment]
@bradwilson Created https://github.com/xunit/xunit.analyzers/pull/171 as a candidate for addressing this issue. Please advise.
